File: 1101-1200/1124-Longest Well-Performing Interval/1124-Longest Well-Performing Interval.py
class Solution:
    def longestWPI(self, hours: List[int]) -> int:
        maxLen = bal = 0
        table = {}
        for i, h in enumerate(hours):
            if hours[i] > 8:
                bal += 1
            else:
                bal -= 1

            table.setdefault(bal, i)
            if bal > 0:
                maxLen = i + 1
            else:                
                if bal - 1 in table:
                    maxLen = max(maxLen, i - table[bal - 1])

        return maxLen

File: 1101-1200/1124-Longest Well-Performing Interval/1124-Longest Well-Performing Interval.cpp
class Solution {
public:
    int longestWPI(vector<int>& hours) {
        int maxLen = 0, bal = 0;
        unordered_map<int, int> table;
        for (int i = 0; i < hours.size(); ++i) {
            if (hours[i] > 8) {
                ++bal;
            }
            else {
                --bal;
            }

            if (!table.count(bal)) {
                table[bal] = i;
            }

            if (bal > 0) {
                maxLen = i + 1;
            }
            else if (table.count(bal - 1)) {
                maxLen = max(maxLen, i - table[bal - 1]);
            }
        }

        return maxLen;
    }
};
